airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigPersistenceFactory.java

```diff
@@ -0,0 +1,151 @@
+   *        has multiple callers, and we want to setup the database only once to prevent race
+   *        conditions.
+   */
+  private ConfigPersistenceFactory(Configs configs, boolean setupDatabase, boolean useConfigDatabase) {
+    this.configs = configs;
+    this.setupDatabase = setupDatabase;
+    this.useConfigDatabase = useConfigDatabase;
+  }
+
+  public static Builder build(Configs configs) {
+    return new Builder(configs);
+  }
+
+  public static class Builder {
+
+    private final Configs configs;
+    private boolean setupDatabase = false;
+    private boolean useConfigDatabase = true;
+
+    private Builder(Configs configs) {
+      this.configs = configs;
+    }
+
+    public Builder setupDatabase(boolean setupDatabase) {
+      this.setupDatabase = setupDatabase;
+      return this;
+    }
+
+    public Builder useConfigDatabase(boolean useConfigDatabase) {
+      this.useConfigDatabase = useConfigDatabase;
+      return this;
+    }
+
+    public ConfigPersistenceFactory get() {
+      return new ConfigPersistenceFactory(configs, setupDatabase, useConfigDatabase);
+    }
+
+  }
+
+  /**
+   * Create a config persistence based on the configs.
+   * <p/>
+   * If config root is defined, create a database config persistence and copy the configs from the
+   * file-based config persistence. Otherwise, seed the database from the yaml files.
+   */
+  public ConfigPersistence create() throws IOException {
+    if (!useConfigDatabase) {
+      Path configRoot = configs.getConfigRoot();
+      LOGGER.info("Use file system config persistence (root: {})", configRoot);
+      return FileSystemConfigPersistence.createWithValidation(configRoot);
+    }
+
+    if (configs.getConfigRoot() == null) {
+      // This branch will only be true in a future Airbyte version, in which
+      // the config root is no longer required and everything lives in the database.
+      return createDbPersistenceWithYamlSeed();
+    }
+
+    return createDbPersistenceWithFileSeed();
+  }
+
+  ConfigPersistence createDbPersistenceWithYamlSeed() throws IOException {
+    ConfigPersistence seedConfigPersistence = new YamlSeedConfigPersistence();
+    return createDbPersistence(seedConfigPersistence);
+  }
+
+  ConfigPersistence createDbPersistenceWithFileSeed() throws IOException {
+    Path configRoot = configs.getConfigRoot();
+    ConfigPersistence fsConfigPersistence = FileSystemConfigPersistence.createWithValidation(configRoot);
+    return createDbPersistence(fsConfigPersistence);
+  }
+
+  ConfigPersistence createDbPersistence(ConfigPersistence seedConfigPersistence) throws IOException {
+    LOGGER.info("Use database config persistence.");
+
+    // When setupDatabase is true, it means the database will be initialized after we
+    // connect to the database. So the database itself is considered ready as long as
+    // the connection is alive. Otherwise, the database is expected to have full data.
+    Function<Database, Boolean> isReady = setupDatabase
```

Comment:
it took me kinda a long time to reason about this. I wonder if it would be easier to understand if we just have two methods in databases `createPostgresDatabaseWithRetry` and `createPostgresDatabaseWithRetryBlockUntilDatabaseIsSeeded`. The latter is obviously a trash name but illustrates what I am trying to clarify.

Reply:
The issue is that the queries we need to run to check if job database and config database are ready are different. So creating two methods in `Databases` won't solve this problem. I can try to simplify the method inside the `ConfigPersistenceBuilder`.